narwhal/node/src/worker_node.rs

```diff
@@ -0,0 +1,329 @@
+        } else {
+            // create a new registry
+            let registry = Registry::new();
+
+            (initialise_metrics(&registry), Some(registry))
+        };
+
+        let mut tx_shutdown = PreSubscribedBroadcastSender::new(NUM_SHUTDOWN_RECEIVERS);
+
+        let handles = Worker::spawn(
+            primary_name,
+            network_keypair,
+            self.id,
+            committee.clone(),
+            worker_cache.clone(),
+            self.parameters.clone(),
+            tx_validator.clone(),
+            store.batch_store.clone(),
+            metrics,
+            &mut tx_shutdown,
+        );
+
+        // store the registry
+        if let Some(registry) = registry {
+            self.swap_registry(Some(registry));
+        }
+
+        // now keep the handlers
+        self.handles.clear();
+        self.handles.extend(handles);
+        self.tx_shutdown = Some(tx_shutdown);
+
+        Ok(())
+    }
+
+    // Will shutdown the worker node and wait until the node has shutdown by waiting on the
+    // underlying components handles. If the node was not already running then the
+    // method will return immediately.
+    async fn shutdown(&mut self) {
+        if !self.is_running().await {
+            return;
+        }
+
+        if let Some(tx_shutdown) = self.tx_shutdown.as_ref() {
+            tx_shutdown
+                .send()
+                .expect("Couldn't send the shutdown signal to downstream components");
+            self.tx_shutdown = None;
+        }
+
+        // Now wait until handles have been completed
+        try_join_all(&mut self.handles).await.unwrap();
+
+        self.swap_registry(None);
+
+        info!("Narwhal worker {} shutdown is complete", self.id);
+    }
+
+    // If any of the underlying handles haven't still finished, then this method will return
+    // true, otherwise false will returned instead.
+    async fn is_running(&self) -> bool {
+        self.handles.iter().any(|h| !h.is_finished())
+    }
+
+    // Helper method useful to wait on the execution of the primary node
+    async fn wait(&mut self) {
+        try_join_all(&mut self.handles).await.unwrap();
+    }
+
+    // Accepts an Option registry. If it's Some, then the new registry will be added in the
+    // registry service and the registry_id will be updated. Also, any previous registry will
+    // be removed. If None is passed, then the registry_id is updated to None and any old
+    // registry is removed from the RegistryService.
+    fn swap_registry(&mut self, registry: Option<Registry>) {
+        if let Some((registry_id, _registry)) = self.registry.as_ref() {
+            self.registry_service.remove(*registry_id);
+        }
+
+        if let Some(registry) = registry {
+            self.registry = Some((self.registry_service.add(registry.clone()), registry));
+        } else {
+            self.registry = None
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct WorkerNode {
```

Comment:
The `WorkerNode` (similarly to the `PrimaryNode`) structs are wrapper/shell structs around the "Inner" ones to allow us modify the internals using a sharable across threads struct and not restricting us from having to use an exclusive `&mut` object in our code.

Reply:
In what way do we intend to modify the internals?